server/task.go

```diff
@@ -0,0 +1,277 @@
+// Nothing is submitted to etcd.
+func newTask(req *api.TaskRequest) *Task {
+	id := &api.TaskID{uuid.NewV4().String()}
+
+	return &Task{key: taskKey(id), Task: &pb.Task{Request: req, Id: id}}
+}
+
+// getTask retrieves a Task from etcd.
+func getTask(ctx context.Context, client *clientv3.Client, id *api.TaskID) (*Task, error) {
+	key := taskKey(id)
+
+	resp, err := client.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+
+	// We're not searching for a range or prefix
+	if len(resp.Kvs) != 1 {
+		return nil, fmt.Errorf("Too many matching keys, found %d", resp.Count)
+	}
+
+	task := &Task{version: resp.Kvs[0].Version, key: key, Task: &pb.Task{}}
+	if err := proto.Unmarshal([]byte(resp.Kvs[0].Value), task.Task); err != nil {
+		return nil, err
+	}
+
+	// Ensure what we requested and what we got back match up
+	if task.Id.Uuid != id.Uuid {
+		return nil, fmt.Errorf("Requested task %s, received %s", id.Uuid, task.Id.Uuid)
+	}
+
+	return task, nil
+}
+
+// taskKey returns the etcd key for a TaskID
+func taskKey(id *api.TaskID) string {
+	return idKey(taskPrefix, id)
+}
+
+// queuedPrefix returns the status key prefix for queued tasks
+func queuedPrefix() string {
+	return queuedPrefixFmt
+}
+
+// runningPrefix returns the status key prefix for tasks running on id
+func runningPrefix(id *api.NodeID) string {
+	return fmt.Sprintf(runningPrefixFmt, id.Uuid)
+}
+
+// completePrefix returns the status key prefix for tasks completed at age
+func completePrefix(age int64) string {
+	return fmt.Sprintf(completePrefixFmt, age)
+}
+
+// canceledPrefix returns the status key prefix for tasks canceledat age
+func canceledPrefix(age int64) string {
+	return fmt.Sprintf(canceledPrefixFmt, age)
+}
+
+// statusKey returns the etcd status key of a Task for a given TaskStatus
+func (t *Task) statusKey(status *api.TaskStatus) string {
+	var prefix string
+
+	switch status.Status.(type) {
+	case *api.TaskStatus_Queued_:
+		prefix = queuedPrefix()
+	case *api.TaskStatus_Running_:
+		prefix = runningPrefix(status.GetRunning().NodeId)
+	case *api.TaskStatus_Complete_:
+		prefix = completePrefix(status.GetComplete().Epoch)
+	case *api.TaskStatus_Canceled_:
+		prefix = canceledPrefix(status.GetCanceled().Epoch)
+	default:
+		// TODO - Is this wise?
+		panic("Unexpected Task status")
+	}
+
+	return idKey(prefix, t.Id)
+}
+
+// taskID converts a status / task key, to a TaskID
+func taskID(key string) *api.TaskID {
+	s := strings.Split(key, "/")
+	return &api.TaskID{Uuid: s[len(s)-1]}
+}
+
+// idKey converts a key prefix to full status / task key
+func idKey(prefix string, key *api.TaskID) string {
+	return fmt.Sprintf(prefix+"%s", key.Uuid)
+}
+
+// setStatus Updates the status of a Task, and updates the Task and its status key in etcd
+func (t *Task) setStatus(ctx context.Context, client *clientv3.Client, newStatus *api.TaskStatus) (err error) {
+	// TODO - Check oldStatus != newStatus. Not sure how to do this without reflection...
```

Comment:
You don't need reflection - just a couple of helper methods. You can even extend the protobuf types with `Compare` if you place them in `api` package.